ERCS/erc-7909.md

```diff
@@ -0,0 +1,511 @@
+
+There are existing ERC proposals that suggest a lockable version of tokens. However, those might not be applicable to existing tokens and also each of them is still individual token contracts, the interoperability is still limited. In this proposal, we are defining a wrapper contract that can wrap any token into a "margin token" and can be used across multiple DEX as collaterals. 
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Margin Token Wrapper Contract
+
+The `MarginTokenWrapper` contract will be an [ERC-1155](./eip-1155.md) contract implementing these core functionalities:
+- **GetTokenId**: [ERC-1155](./eip-1155.md) requires each token to have an `uint256` id, 
+  - For [ERC-20](./eip-20.md), we will just use `uint256(uint160(tokenAddress))` as id, in which `tokenAddress` is the token's smart contract address, so that it is guaranteed to be unique. 
+  - For [ERC-721](./eip-721.md) and [ERC-1155](./eip-1155.md), we can use `tokenAddress || tokenId` as id.
+- **Mint**: The `owner` needs to approve the contract to `transferFrom`, when minting, the contract will pull tokens from `sender`'s balance and mint the corresponding wrapped token with the same amount to the `owner`.
+- **ApproveLock**: To use the wrapped token as margin, the `owner` needs to approve another address, marked as `locker`, to lock assets, the lock approval will come with
+  - `tokenAddress` or `id`: only 1 type of wrapped token is approved
+  - `limit`: the maximum amount the locker can lock
+  - `expiration`: a timestamp when the lock will be automatically released, this is refreshed everytime a new approval is effective
+- **PermitLock**: A signature version of **ApproveLock**, following 712 sign typed data standard.
+- **Lock**: Once locking is approved, the `locker` can lock a specific amount of the `owner` of the specific token. Lock operation will consume the `limit` defined in the approval.
+- **Unlock**: `locker` can unlock anytime
+- **Liquidate**: Liquidation is transferring locked token, there are 2 scenarios:
+  - `locker` can transfer the locked assets anytime
+  - `owner` can transfer the locked assets if the `expiration` of a locker is reached
+
+The following functions will only be applicable to unlocked balances:
+- **Approve**: the same as common implementation
+- **Transfer**: the same as common implementation but only able to transfer unlocked balances
+- **TransferFrom**: the same as common implementation
+- **Burn**: the `owner` can always burn any unlocked balances to get the corresponding unwrapped token back
```

Comment:
I'd recommend providing a Solidity interface. It's much more strict than English text.